src/fs/metadata_log.cc

```diff
@@ -101,18 +101,50 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
     } else {
         cut_out_data_range(file, new_data_vec.data_range);
     }
-
     file.data.emplace(new_data_vec.data_range.beg, std::move(new_data_vec));
 }
 
 void metadata_log::cut_out_data_range(inode_info::file& file, file_range range) {
-    file.cut_out_data_range(range, [](inode_data_vec data_vec) {
-        (void)data_vec; // TODO: for compaction: update used inode_data_vec
+    using CR = inode_info::file::cut_result;
+    file.cut_out_data_range(range, [&](inode_data_vec data_vec, CR cr) {
+        int new_writes_cnt;
+        if (cr == CR::BOTH) {
+            new_writes_cnt = 2;
+        } else if (cr == CR::LEFT || cr == CR::RIGHT) {
+            new_writes_cnt = 1;
+        } else {
+            new_writes_cnt = 0;
+        }
+        std::visit(overloaded {
+            [&](inode_data_vec::in_mem_data& mem) {
+                size_t log_decrease = ondisk_entry_size<ondisk_small_write_header>(data_vec.data_range.size());
+                _rewrite_log_size -= log_decrease;
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_small_write_header>(0);
+            },
+            [&](inode_data_vec::on_disk_data& disk_data) {
+                // TODO: Differentiate between large write and large write without mtime
+                size_t log_decrease = data_vec.data_range.size() == _cluster_size ?
+                    ondisk_entry_size<ondisk_large_write>() :
+                    ondisk_entry_size<ondisk_medium_write>();
+                _rewrite_log_size -= log_decrease;
+                // Remaining data is smaller than original so it's less than large write
+                _rewrite_log_size += new_writes_cnt * ondisk_entry_size<ondisk_medium_write>();
+                _clusters[offset_to_cluster_id(disk_data.device_offset, _cluster_size)].remove_data(disk_data.device_offset, data_vec.data_range);
+            },
+            [&](inode_data_vec::hole_data&) {
+            },
+        }, data_vec.data_location);
     });
 }
 
 inode_info& metadata_log::memory_only_create_inode(inode_t inode, bool is_directory, unix_metadata metadata) {
     assert(_inodes.count(inode) == 0);
+    size_t log_increase = ondisk_entry_size<ondisk_create_inode>();
+    _ondisk_log_size += log_increase;
```

Comment:
_ondisk_log_size is redundant (too high accuracy, maybe even counting not what we want) imo. What is enough is the number of clusters that metadata log takes.

Reply:
If you want to stay with _ondisk_log_size, does it accounts for added padding when flushing metadata log?

Reply:
We can't calculate precisely the number of clusters in the compacted metadata log nor account for added padding in it, so for the sake of comparison we're not doing it for the written metadata log too.

Reply:
Yeah, that is true (but I think it is achievable although hard to take padding into account, regarding _ondisk_log_size).
Anyway, calculating the number of clusters that metadata log takes is trivial, this takes padding into account and is accurate (because this is the space that metadata log occupies). Padding is problematic to foresee -- I totally agree. But that is the margin of error that should be negligible.
_ondisk_log_size is not a good comparison because scenarios like this exist: suppose that I fill the log with 10000 entries of creating inode (that is really a small entry) and after adding each, I trigger the flush_log(), so that the padding is added. This way metadata log will take about 100 times more space than _ondisk_log_size is indicating (this number is for alignment == 4096). This is really bad that, because we could reclaim a lot of space using compaction, but this metric would indicate that it is not worth compacting because _ondisk_log_size == _rewrite_log_size (more or less).